### PR TITLE
Feature/add hash query builder

### PR DESCRIPTION
**Pull request by ravit-db**
- Added Hash Query Builder for Data Comparison
- Added Hash Algo to pick the algorithm at runtime depending upon the dialect

### CODE REVIEW

File: src/databricks/labs/remorph/reconcile/query_builder/hash_query.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        res = (
+            exp.select(*hash_col_with_transform + key_cols_with_transform)
+            .from_(":tbl")
+            .where(self.filter)
+            .sql(dialect=self.source)
+        )
+
+        return res
+
+    def _add_transformations(self, aliases: list[exp.Alias], source: str) -> list[exp.Alias]:
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/query_builder/hash_query.py:36]: This method looks generic enough to move to `QueryBuilder` class, thoughts?

File: tests/unit/reconcile/query_builder/test_hash_query.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+from databricks.labs.remorph.reconcile.query_builder.hash_query import HashQueryBuilder
+from databricks.labs.remorph.reconcile.recon_config import Filters
+
+
+def test_hash_query_builder_for_snowflake_src(table_conf_with_opts, schema):
```
Comment (sundarshankar89) [inline tests/unit/reconcile/query_builder/test_hash_query.py:5]: missing tests for databricks

File: src/databricks/labs/remorph/reconcile/recon_config.py [hunk 3]
```diff
@@ -165,3 +168,9 @@ class ReconcileOutput:
     missing_in_src: DataFrame
     missing_in_tgt: DataFrame
     mismatch: DataFrame | None = None
+
+
+@dataclass
+class HashAlgoMap:
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/recon_config.py:174]:
```suggestion
class DialectHashConfig:
    dialect: str
    hash_algorithm: list[Callable]
```

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/reconcile/query_builder/hash_query.py (hunk 1 above)
Comment (ravit-db) [reply src/databricks/labs/remorph/reconcile/query_builder/hash_query.py:36]: Done.

File: tests/unit/reconcile/query_builder/test_hash_query.py (hunk 2 above)
Comment (ravit-db) [reply tests/unit/reconcile/query_builder/test_hash_query.py:5]: Added.

**Review by sundarshankar89 — APPROVED** [review]
LGTM

**Pull request merged**